### ISSUE TITLE
proposal: Rename "implicit cast" to "type coercion"

### ISSUE THREAD

**Issue by Tetralux** [ziglang/zig#3361]
This is fresh of the assembly line that is my brain, so it's yet to be fully hashed out in my mind at the time of opening, but...

We currently call things "implicit casts" despite explicitly providing a type to cast to.
In IRC, I suggested that things like `var x: i32 = not_an_i32` is actually a type coercion and not a cast.
@andrewrk suggested I open this proposal for discussion etc, so here we are.

There's some discussion to be had here, but roughly, the synopsis goes something like this -- more to follow in the comments, no doubt:

```zig
var x: i32 = not_an_i32
```
This is called an implicit cast; but `type coercion` would fit better since the syntax doesn't really imply a cast here. You know what I'm getting at here if you compare it to the syntax for normal casts.

```zig
i32(x)
```
This is also currently an implicit cast, but yet the syntax implies you are explicitly casting; again, `type coercion`; same as the above.
As an aside, I'm inclined to think that this syntax should actually behave the same as `@intCast`, and not how it does currently but, we currently use this to give a type to an otherwise "untyped" value - so... *shrugs*.
Point is, this is definitely **not** an implicit cast.

```zig
@ptrToInt(x)
```
An explicit cast; equivalent to `@intCast(usize, x)` - except that for clarity and to make it harder to write errors, we have this special cast instead -- it's still an example of an explicit cast though.

TL; DR: `var x: i32 = not_an_i32` and `i32(x)` behave the same way, but are both currently called "implicit casts", despite how the first one doesn't seem like a cast at all, and how the type is even explicitly provided in the second one.
Renaming this functionality to `type coercion` fits this better.

**Comment by andrewrk** [comment]
Landed in 6d5abf87ecd3509c6fb8b9c917b73b4db2ae59ff

Thanks for the well-written proposal.
